rct_optimizations/src/validation/noise_qualifier.cpp

```diff
@@ -0,0 +1,132 @@
+#include <Eigen/Dense>
+#include <Eigen/Geometry>
+#include <rct_optimizations/validation/noise_qualifier.h>
+#include <rct_optimizations/experimental/pnp.h>
+
+//accumulators
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics/stats.hpp>
+#include <boost/accumulators/statistics/mean.hpp>
+#include <boost/accumulators/statistics/variance.hpp>
+
+
+
+namespace rct_optimizations
+{
+
+ std::vector<NoiseStatistics> qualifyNoise2D(const std::vector<PnPProblem>& params)
+ {
+
+  std::vector<NoiseStatistics> output;
+  output.reserve(6);
+
+  std::vector<Eigen::Isometry3d> solution_transforms;
+
+  using namespace boost::accumulators;
+  accumulator_set<double, stats<tag::mean, tag::variance>> x_acc;
+  accumulator_set<double, stats<tag::mean, tag::variance>> y_acc;
+  accumulator_set<double, stats<tag::mean, tag::variance>> z_acc;
+  accumulator_set<double, stats<tag::mean, tag::variance>> r_acc;
+  accumulator_set<double, stats<tag::mean, tag::variance>> p_acc;
+  accumulator_set<double, stats<tag::mean, tag::variance>> yw_acc;
+
+
+  for (auto& prob : params)
+  {
+    rct_optimizations::PnPResult result;
+
+    result = rct_optimizations::optimize(prob);
+
+    if (result.converged)
+    {
+      //we will save the full result here for debugging purposes
+      solution_transforms.push_back(result.camera_to_target);
+
+      x_acc(result.camera_to_target.translation()(0));
+      y_acc(result.camera_to_target.translation()(1));
+      z_acc(result.camera_to_target.translation()(2));
+
+      Eigen::Vector3d ea = result.camera_to_target.linear().eulerAngles(0,1,2);
+      r_acc(ea(0));
+      p_acc(ea(1));
+      yw_acc(ea(2));
```

Comment:
I think I am misunderstanding this, but aren't the results of decomposing a rotation matrix into x,y,z components analogous to Euler angles? Or at least susceptible to the same non-uniqueness (ie flipping pi about the x-axis is functionally the same as pi about z + pi about y)

Reply:
I think you're right about still being subject to non-uniqueness. We just need to get the orientation in a form where you can compare them in your unit tests. Currently you note:
> Euler angles are excluded because of singularities

Angle axis representation (or quaternions for that matter) would be an alternative for getting around this issue